src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -1,47 +1,145 @@
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
+        if (index == size) {
+            add(value);
+            return;
+        }
+        if (index == 0) {
+            Node<T> newNode = new Node<>(null, value, head);
+            head.prev = newNode;
+            head = newNode;
+            size++;
+        } else {
+            Node<T> currentNode = getNodeByIndex(index);
+            Node<T> newNode = new Node<>(currentNode.prev, value, currentNode);
+            newNode.prev = currentNode.prev;
+            currentNode.prev.next = newNode;
+            currentNode.prev = newNode;
+            newNode.next = currentNode;
+            size++;
+        }
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T t : list) {
+            add(t);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        Node<T> currentNode = getNodeByIndex(index);
+        return currentNode.element;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        Node<T> currentNode = getNodeByIndex(index);
+        T oldValue = currentNode.element;
+        currentNode.element = value;
+        return oldValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        Node<T> currentNode = getNodeByIndex(index);
+        removeNode(currentNode);
+        return currentNode.element;
     }
 
     @Override
     public boolean remove(T object) {
+        Node<T> node = head;
+        for (int i = 0; i < size; i++) {
+            if (Objects.equals(node.element, object)) {
+                remove(i);
+                return true;
+            }
+            node = node.next;
+        }
         return false;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private Node<T> getNodeByIndex(int index) {
+        checkValidationByIndex(index);
+        Node<T> currentNode;
+        if (index < size / 2) {
+            currentNode = head;
+            for (int i = 0; i < index; i++) {
+                currentNode = currentNode.next;
+            }
+        } else {
+            currentNode = tail;
+            for (int i = size - 1; i > index; i--) {
+                currentNode = currentNode.prev;
+            }
+        }
+        return currentNode;
+    }
+
+    private void removeNode(Node<T> currentNode) {
+        if (currentNode.prev == null && currentNode.next == null) {
+            head = tail = null;
+        }
+        if (currentNode.prev == null) {
+            head = currentNode.next;
+        } else if (currentNode.next == null) {
+            tail = currentNode.prev;
+        } else {
+            currentNode.prev.next = currentNode.next;
+            currentNode.next.prev = currentNode.prev;
+        }
+        size--;
```

Comment:
```suggestion
        if (currentNode.prev == null) {
            head = currentNode.next;
        } else {
            currentNode.prev.next = currentNode.next;
        }
        if (currentNode.next == null) {
            tail = currentNode.prev;
        } else {
            currentNode.next.prev = currentNode.prev;
        }
        size--;
```

we could simplify this